### ISSUE TITLE
Translate aggregate functions without GroupBy

### ISSUE THREAD

**Issue by roji** [dotnet/efcore#29200]
Max can be used over a collection navigation in a projection:

```c#
_ = await ctx.Authors
    .Select(a => new { Author = a, Books = a.Books.Max(b => b.Id) })
    .ToArrayAsync();
```

```sql
SELECT [a].[Id], [a].[Name], (
    SELECT MAX([b].[Id])
    FROM [Books] AS [b]
    WHERE [a].[Id] = [b].[AuthorId]) AS [Books]
FROM [Authors] AS [a]
```

However, the same can't be done with string.Join:

```c#
_ = await ctx.Authors
    .Select(a => new { Author = a, Books = string.Join(", ", a.Books.Select(b => b.Name)) })
    .ToArrayAsync();
```

(this causes client evaluation)

The query above can be rewritten using a GroupBy, which does work:

```c#
_ = await ctx.Books
    .GroupBy(b => b.Author)
    .Select(g => new { Author = g.Key, Books = string.Join(", ", g.Select(b => b.Name)) })
    .ToArrayAsync();
```

**Comment by ajcvickers** [comment]
Also, just attempting to use the aggregate function is enough to force the switch to client evaluation (and fail) even if client eval would not be used without the aggregate. For example:

```C#
var query = context.Authors
    .Select(author => new
    {
        Author = author,
        TotalCost = author.Books.Sum(book => book.Detail.Price),
        // VariancePopulation = EF.Functions.VariancePopulation(author.Books.Select(post => post.Detail.Price)),
    });
```

Generates:

```sql
      SELECT [a].[Id], [a].[Name], (
          SELECT COALESCE(SUM([b].[Detail_Price]), 0)
          FROM [Books] AS [b]
          WHERE [a].[Id] = [b].[AuthorId]) AS [TotalCost]
      FROM [Authors] AS [a]
```

But uncommenting the function call causes the client-eval warning.

**Comment by roji** [comment]
> Also, just attempting to use the aggregate function is enough to force the switch to client evaluation (and fail) even if client eval would not be used without the aggregate. For example:

Right, I think that's expected given there's a translation failure with the custom aggregates (regular switch to client eval on failure in top projection).

**Comment by ajcvickers** [comment]
But why is there a translation failure?

**Comment by roji** [comment]
That's what this issue is about :) Just saying that the "transparent" switch to client eval here is expected.

**Comment by ajcvickers** [comment]
Surely this is a bug then?

**Comment by roji** [comment]
Yes, that's why I opened this issue ;)

**Comment by ajcvickers** [comment]
But you labeled is as an enhancement...

**Comment by roji** [comment]
Or I mean, you could consider it an as-of-yet unsupported query shape (so enhancement). Doesn't really matter...

**Comment by ajcvickers** [comment]
It matters if we are going to fix it for 7.0. :-)

**Comment by ajcvickers** [comment]
Note from triage: this is a limitation based on the aggregate function having only an `IEnumerable` overload, as opposed to methods like `Max` which also have an `IQueryable` overload and are there for handled differently in the expression tree. We will document this limitation for 7.0.

**Comment by roji** [comment]
Another instance: https://github.com/npgsql/efcore.pg/issues/2569

**Comment by aradalvand** [comment]
Would be nice if this got implemented in EF8, seems like one of those things that's pretty basic, but currently lacking.
